### PR TITLE
VSF should be flexible to support id values to be numeric as well as alphanumeric

### LINKED ISSUE #3456: VSF should be flexible to support id values to be numeric as well as alphanumeric
## What is the motivation for adding / enhancing this feature?
We have all the id values of product, category, attribute etc defined as numeric type.
https://github.com/DivanteLtd/vue-storefront/blob/master/core/modules/catalog/types/Product.ts
https://github.com/DivanteLtd/vue-storefront-api/blob/master/config/elastic.schema.product.json
https://github.com/DivanteLtd/vue-storefront-integration-boilerplate/blob/master/1.%20Expose%20the%20API%20endpoints%20required%20by%20VS/Product%20Attributes.md
https://github.com/DivanteLtd/vue-storefront-integration-boilerplate/blob/master/1.%20Expose%20the%20API%20endpoints%20required%20by%20VS/Required%20API%20specification.md
https://github.com/DivanteLtd/vue-storefront-api/blob/master/config/elastic.schema.attribute.json
It may be possible that the ecommerce ERP, it is integrated with may have support for alphanumeric identifier.
We would like to contribute towards the effort to make it support alphanumeric if community finds it useful.
Slack thread link: https://vuestorefront.slack.com/archives/CA5B80T4M/p1567406131004600

## What are the acceptance criteria 
Enable VSF to support **id** values to accept alphanumeric as well as numeric values. 
- [ ] ...

## Can you complete this feature request by yourself?

- [x] YES
- [ ] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [x] This is a normal feature request. This should be available on https://test.storefrontcloud.io and then after tests this can be added to next Vue Storefront version. In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] (**Pick this option only if you're sure**) This is an important improvement request for current Release Candidate version on https://next.storefrontcloud.io and should be placed in next RC version. In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] (**Pick this option only if you're sure**) This is a critical improvement request for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version. In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

### PR DESCRIPTION
**Pull request by adityasharma7**
### Related issues

closes #3456

### Short description and why it's useful

Though the typescript files types/X.ts do not force any data to be numeric, we can make values to support any type.

src/themes/default/components/core/blocks/Reviews/Reviews.vue
There was a warning on console changing id to string

[Vue warn]: Invalid prop: type check failed for prop "productId". Expected Number with value 1, got String with value "1".

found in

---> <Reviews> at src/themes/default/components/core/blocks/Reviews/Reviews.vue
       <Product> at src/themes/default/pages/Product.vue
         <DefaultLayout> at src/themes/default/layouts/Default.vue
           <App> at src/themes/default/App.vue
             <Root> 

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [ ] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [ ] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and I am aware that every new feature should be a module

### CODE REVIEW

File: core/data-resolver/types/DataResolver.d.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ import Review from 'core/modules/review/types/Review';
 declare namespace DataResolver {
 
   interface CategorySearchOptions {
-    parentId?: number,
+    parentId?: any,
```
Comment (lukeromanowicz) [inline core/data-resolver/types/DataResolver.d.ts]: In this case, these types should be number|string rather than any.

**Comment by adityasharma7** [comment]
Thanks @patzick for the review. By changing it to any, I intend to make it accept a string as well as int. It seems the primary integration VSF has is with Magento which gives these values as numeric, we don't want to break the existing system.

File: core/data-resolver/types/DataResolver.d.ts (hunk 1 above)
Comment (adityasharma7) [reply core/data-resolver/types/DataResolver.d.ts]: Makes sense @lukeromanowicz. These changes can be applied to all the interface files. WDYT?

**Comment by pkarw** [comment]
It should be number OR string as you changed - I mean, this is just my opinion :)

File: core/data-resolver/types/DataResolver.d.ts (hunk 1 above)
Comment (adityasharma7) [reply core/data-resolver/types/DataResolver.d.ts]: I have updated the PR. Please check.

**Comment by adityasharma7** [comment]
Thanks @pkarw for your inputs :)
@patzick I have updated the PR. Please check.

**Comment by adityasharma7** [comment]
It seems in PR template 
**TypeScript Action Plan** and **Vue Storefront Modules** under **Contribution and currently important rules acceptance** are broken.
It should be
https://github.com/DivanteLtd/vue-storefront/blob/master/docs/guide/basics/typescript.md
https://github.com/DivanteLtd/vue-storefront/blob/master/docs/guide/modules/introduction.md

**Comment by pkarw** [comment]
@ArturDivante please test it on the branch we shouldn't have got any issues in the console; we also need to test the server cart sync in all possible scenarios (logged in/non-logged in user etc)

File: src/themes/default/components/core/blocks/Reviews/Reviews.vue [hunk 2]
```diff
@@ -126,7 +126,7 @@ export default {
   },
   props: {
     productId: {
-      type: Number,
+      type: null,
```
Comment (patzick) [inline src/themes/default/components/core/blocks/Reviews/Reviews.vue]: Just one more here - you can also do here Number or String in vue prop :)

Comment (adityasharma7) [reply src/themes/default/components/core/blocks/Reviews/Reviews.vue]: Updated. Please check. 
Updated changelog too

**Comment by pkarw** [comment]
@adityasharma7 doesn't it require any changes to `vue-storefront-api/config/*.schema.json` files?

**Comment by adityasharma7** [comment]
@pkarw This change certainly will not break that. 
We don't use VSF-api but slightly modified schema files. 
I think either we can make the type of those fields as a string or numeric else we will have to remove the schema for id fields. If we remove schema, elastic search allows data pushed of any type. I am not sure though. I will dig deeper.

**Comment by adityasharma7** [comment]
I did some more exploration and found [Mutate Filter plugin](https://www.elastic.co/guide/en/logstash/5.6/plugins-filters-mutate.html#plugins-filters-mutate-convert) which allows conversion of fields type.
We can certainly try it with [elasticdump](https://www.npmjs.com/package/elasticdump).
I have not tested it yet as I am bit occupied. I think it will need some grain attention. I will try to revert on this asap.

**Comment by alinadivante** [comment]
@pkarw is it possible that this PR causes a problem with instant checkout?
image

**Comment by pkarw** [comment]
I think it's something different - no ide why `order/placeOrder` is missing - @andrzejewsky can you please take a look?

**Comment by alinadivante** [comment]
It works on the master branch

**Pull request merged**